compiler/resolution/lowerIterators.cpp

```diff
@@ -33,6 +33,31 @@
 #include "symbol.h"
 
 
+// Get the corresponding iterator FnSymbol, insead of:
+//   ...->defaultInitializer->getFormal(1)->type->defaultInitializer
+
+FnSymbol* getTheIteratorFn(Symbol* ic) {
+  return getTheIteratorFn(ic->type);
+}
+
+FnSymbol* getTheIteratorFn(CallExpr* call) {
+  return getTheIteratorFn(call->get(1)->typeInfo());
+}
+
+FnSymbol* getTheIteratorFn(Type* icType) {
+  // either an IC or a tuple thereof
+  // the asserts document the current state
+  bool gotTuple = icType->symbol->hasFlag(FLAG_TUPLE);
+  INT_ASSERT(gotTuple || icType->symbol->hasFlag(FLAG_ITERATOR_CLASS));
+  Type* irType = icType->defaultInitializer->getFormal(1)->type;
```

Comment:
Now you have space to explain that we stash the getIterator function in the defaultInitializer field of an IC type, and that the first argument of getIterator is always an IR type.
Then explain that we stash the original iterator in the defaultInitializer field of the IR type.
Explain that this is very bad practice and that we should instead subclass the type symbol for both iterator classes and iterator records (separately), so that we have a "proper" place to hang such information.